crates/wasm-mutate/src/mutators/peephole/mod.rs

```diff
@@ -0,0 +1,252 @@
+
+use rand::{
+    prelude::{IteratorRandom, SliceRandom, SmallRng},
+    Rng,
+};
+use wasm_encoder::{CodeSection, Function, Module};
+use wasmparser::{BinaryReaderError, CodeSectionReader, FunctionBody, Operator};
+
+use crate::{
+    mutators::peephole::swap_commutative::SwapCommutativeOperator, ModuleInfo, Result, WasmMutate,
+};
+
+use super::Mutator;
+
+pub mod swap_commutative;
+
+pub struct PeepholeMutator;
+
+/// Meta mutator for peephole
+impl Mutator for PeepholeMutator {
+    fn mutate(
+        &self,
+        config: &crate::WasmMutate,
+        rnd: &mut rand::prelude::SmallRng,
+        info: &mut crate::ModuleInfo,
+    ) -> Result<Module> {
+        // Parse the module to get opcodes
+        let mut codes = CodeSection::new();
+        let code_section = info.get_code_section();
+        let mut sectionreader = CodeSectionReader::new(code_section.data, 0)?;
+        let function_count = sectionreader.get_count();
+
+        let peep_optimizers: &Vec<Box<dyn CodeMutator>> = &vec![Box::new(SwapCommutativeOperator)];
+
+        // Split where to start looking for mutable function
+        let function_to_mutate = rnd.gen_range(0, function_count);
+        let all_readers = (0..function_count)
+            .map(|fidx| sectionreader.read().unwrap())
+            .collect::<Vec<FunctionBody>>();
+
+        // Since we can have several positions for the same mutator it is better to group them by mutator reference
+        let mut applicable: HashMap<String, Vec<(usize, usize, &Box<dyn CodeMutator>)>> =
+            HashMap::new();
+
+        (function_to_mutate..function_count)
+            .chain(0..function_to_mutate)
+            .fold(&mut applicable, |prev, fidx| {
+                let reader = all_readers[fidx as usize];
+                let operatorreader = reader.get_operators_reader().unwrap();
+                let operators = &operatorreader
+                    .into_iter()
+                    .collect::<wasmparser::Result<Vec<Operator>>>()
+                    .unwrap();
+                let operatorscount = operators.len();
+
+                let opcode_to_mutate = rnd.gen_range(0, operatorscount);
+                (opcode_to_mutate..operatorscount)
+                    .chain(0..opcode_to_mutate)
+                    .fold(prev, |innerprev, idx| {
+                        for peephole in peep_optimizers {
+                            if peephole.can_mutate(config, &operators, idx).unwrap() {
+                                // We can have several mutators, lets group by mutator
+                                // TODO, find better key ?
+                                innerprev
+                                    .entry(peephole.name())
+                                    .or_insert(Vec::new())
+                                    .push((fidx as usize, idx, peephole));
+                            }
+                        }
+                        innerprev
+                    })
+            });
```

Comment:
I think as soon as we find an opcode where we have applicable mutators, we want to choose one of them and apply it there.

Because we are choosing a random opcode in the function to start searching for mutations to apply, and we are still checking all applicable mutators against this opcode and choosing a random one of the applicable ones, this should still result in a random mutation, and shouldn't skew our distribution.

But because we aren't checking every single peephole mutator against every single opcode in the function, this should be much faster for large functions.

Reply:
Improved, now we mutate as soon as we have at least one mutator for a function at an specific operator